### ISSUE TITLE
[java] False positive UnusedPrivateMethod 

### ISSUE THREAD

**Issue by schosin** [pmd/pmd#5145]
**Affects PMD Version:** 
7.3.0

**Rule:**

Please provide the rule name and a link to the rule documentation:
https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#unusedprivatemethod

**Description:**

False positive for private method used by lambda and recursivly by itself. 

Removing the ternary operator for dao access, replacing the collect on the dao results or removing the recursive calls all remove the false positive.

**Code Sample demonstrating the issue:**

https://github.com/schosin/pmd-unused-private-method-false-positive

**Expected outcome:**

PMD reports a violation for method `falsePositive`, but that's wrong. That's a false positive.

**Running PMD through:** *Maven*

**Comment by jsotuyod** [comment]
Upon inspection, this is a type resolution issue.

PMD is unable to determine the variable type for `lookup` [here](https://github.com/schosin/pmd-unused-private-method-false-positive/blob/9513f209614348cf1ba9a7396c2780f1183212a6/src/main/java/de/schosin/pmd/UnusedPrivateMethodFalsePositive.java#L22-L24)

```java
        var lookup = System.currentTimeMillis() > 100L // permission based
                ? Dao.source1().collect(groupingByNullable(Data::parent)) // DAO call #1
                : Dao.source2().collect(groupingByNullable(Data::parent)); // DAO call #2
```

Explicitly setting the type `Map<String, List<Data>>` will work as expected.

Moreover, the type of the calls `Dao.source2().collect(groupingByNullable(Data::parent))` is inferred properly if the ternary is removed to change the code to `var lookup = Dao.source2().collect(groupingByNullable(Data::parent))`. It seems the ternary is confusing the type resolver.

**Comment by jsotuyod** [comment]
A little digging into this, the issue seems to be that `PolyResolution.polyTypeOtherCtx` tries to solve this `ASTConditionalExpression` using `ConditionalMirrorImpl`… this class however, resolves invocations expressions differently than `PolyResolution.polyTypeOtherCtx`… under `ConditionalMirrorImpl.standaloneExprTypeInConditional` we simply call `Infer.getCompileTimeDecl`, but as the method is generic, the invocation phase can't be skipped, but is never called (no call is made to `Infer.finishInstantiation`), so the results are never incorporated and the inferred type for the expression is `R` instead of `Map<String, List<Data>>`.

I'm still trying to understand wether…
 * this statement is actually not standalone but a poly expression, and should be treated differently. The JLS says…
 > A method invocation expression is a poly expression if all of the following are true:
 > * The invocation appears in an assignment context or an invocation context ([§5.2](https://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html#jls-5.2), [§5.3](https://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html#jls-5.3)). ✅ 
> * If the invocation is qualified (that is, any form of MethodInvocation except for the first), then the invocation elides TypeArguments to the left of the Identifier. ❓ 
> * The method to be invoked, as determined by the following subsections, is generic ([§8.4.4](https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-8.4.4)) and has a return type that mentions at least one of the method's type parameters.  ✅ 

My issue is with the second line… I'm unsure if the "IF" means that line only applies to qualified invocations (and this statement would be standalone, as the other 2 are met); or it should read as "is qualified AND … " and therefore this statement would not be standalone, for not meeting that requirement. I'll try with a newer version of the JLS later, hoping for clearer wording.

* It doesn't matter, and we simply need to change `ConditionalMirrorImpl.standaloneExprTypeInConditional` to not call `factory.infer.getCompileTimeDecl` but `factory.infer.inferInvocationRecursively` and return the expression type… if it was standalone the invocation phase can be skipped, and no harm is done resolving it; if it's not, we still need to resolve it…
* something else…

@oowekyala any insights?

**Comment by jsotuyod** [comment]
> I'll try with a newer version of the JLS later, hoping for clearer wording.

Unfortunately, the wording is the same under JLS for Java 22 https://docs.oracle.com/javase/specs/jls/se22/html/jls-15.html#jls-15.12

> * It doesn't matter, and we simply need to change `ConditionalMirrorImpl.standaloneExprTypeInConditional` to not call `factory.infer.getCompileTimeDecl` but `factory.infer.inferInvocationRecursively` and return the expression type… if it was standalone the invocation phase can be skipped, and no harm is done resolving it; if it's not, we still need to resolve it…

A very naive test shows this works for this particular case to be successfully resolved.

**Comment by oowekyala** [comment]
> My issue is with the second line… I'm unsure if the "IF" means that line only applies to qualified invocations (and this statement would be poly, as the other 2 are met); or it should read as "is qualified AND … " and therefore this statement would not be poly, for not meeting that requirement. I'll try with a newer version of the JLS later, hoping for clearer wording.

Your first interpretation is correct. This line is used to make method invocations that have explicit type arguments not poly (which means, their invocation type depends only on the explicit type arguments, not on the context). It would not make sense for methods with explicit type arguments to be poly in this context, as the type arguments would not be useful then.

In this code though, although the method invocations *could* be poly, they in fact are not. Poly expressions use the context type as additionnal information during type inference. However here the context is an assignment context with an inferred `var` type. Since the context type is not provided, it cannot be used for type inference, instead it itself needs to be inferred. Therefore the invocations need to be treated as standalone exprs in order to infer the `var` type. This is mentioned in [JLS 14.4.1](https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.4.1) (in newer versions, since Java 8 didn't have type inference for local variables):

> If the LocalVariableType is var, then let T be the type of the initializer expression when treated as if it did not appear in an assignment context, and were thus a standalone expression ([§15.2](https://docs.oracle.com/javase/specs/jls/se17/html/jls-15.html#jls-15.2)). The type of the local variable is the upward projection of T with respect to all synthetic type variables mentioned by T ([§4.10.5](https://docs.oracle.com/javase/specs/jls/se17/html/jls-4.html#jls-4.10.5)).

I haven't looked at where in the code we should correct the logic. I think you are on the right way though. 

I suppose the logic to fetch the context may be at fault here. Maybe you can look at the following code (same but without the conditional):
```
var foo =  Dao.source1().collect(groupingByNullable(Data::parent));
```
If type inference succeeds here, then the steps it takes are those that should be taken in the conditional example.

> but factory.infer.inferInvocationRecursively and return the expression type… if it was standalone the invocation phase can be skipped, and no harm is done resolving it; if it's not, we still need to resolve it…

As I mentioned the expression needs to be treated as standalone so we shouldn't attempt resolution that uses the context type I think.

**Comment by jsotuyod** [comment]
Thanks for the clarification.

> I suppose the logic to fetch the context may be at fault here. Maybe you can look at the following code (same but without the conditional):
> 
> ```
> var foo =  Dao.source1().collect(groupingByNullable(Data::parent));
> ```

Yes, that resolves. As I mentioned, `ConditionalMirrorImpl` is the one at fault here, `PolyResolution.polyTypeOtherCtx` would resolve the individual branch without issue.

> As I mentioned the expression needs to be treated as standalone so we shouldn't attempt resolution that uses the context type I think.

The call site is setting the expected type to `null`, so the context should not be used. On this case, the bounds for the generics on `Stream<Data>.collect` are bound by the generic types in `groupingByNullable` and the method reference to `Data::parent` which are part of the expression itself.

The question here is, provided the call site set the expected type to `null`, is there any reason why in any circumstance we would want to use `getCompileTimeDecl` and get the generic signature return type rather than the actual type of the expression at compile time?

I see 3 paths forward:
- Make `Infer.getCompileTimeDecl` private, and simply use `Infer.inferInvocationRecursively` with the appropriate call site info
  - The resolved type is committed, even upon failures, for all usages of `Infer.getCompileTimeDecl`
- Completely change the behavior of `Infer.getCompileTimeDecl` to actually do what `Infer.goToInvocationWithFallback` does, resolving the generics based on the Call Site info, falling back to removing unresolved `tvars` with `ts.ERROR` but not committing the resolved type onto the expression.
  - As an optimization we may conditionally commit the resolved type only if resolution was successful.
  - It assumes an expression may be resolved as standalone at some point during analysis, but as poly in another, producing different outputs; but we NEVER want the generics if they can be resolved.
- We only change `ConditionalMirrorImpl` to not use `Infer.getCompileTimeDecl`, but a different method to do either of the ones above (this assumes there is some condition where we want the generics even if the expression can be resolved)

I'm very unsure about option 1, options 2 and 3 look saner to me. But I'm unsure which one is correct at this point. Would need to assess all usages of `Infer.getCompileTimeDecl`…

**Comment by oowekyala** [comment]
Ok I looked a bit more into it and I think we don't need to do too drastic changes (I'm preparing a PR)

Remember the conditions:
> A method invocation expression is a poly expression if all of the following are true:
>  1. The invocation appears in an assignment context or an invocation context ([§5.2](https://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html#jls-5.2), [§5.3](https://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html#jls-5.3)).
> 2.  If the invocation is qualified (that is, any form of MethodInvocation except for the first), then the invocation elides TypeArguments to the left of the Identifier.
> 3. The method to be invoked, as determined by the following subsections, is generic ([§8.4.4](https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-8.4.4)) and has a return type that mentions at least one of the method's type parameters.

The corresponding implementation is here:

https://github.com/pmd/pmd/blob/07f96e79ad75341bf017f5090da63e884dd4df35/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/ast/MethodInvocMirror.java#L35-L36

As you can see it actually only checks condition 3 (context dependence), not whether the context is missing, or whether type arguments are present. I think a fix should start by fixing this.
